scripts/metamath-extract.py

```diff
@@ -36,86 +53,151 @@ def compressed_proof_get_lemmas(proof: str) -> tuple[str, ...]:
+        assert isinstance(statement.statements[-1], ProvableStatement)
+        return (
+            cast('tuple[DisjointStatement | EssentialStatement, ...]', tuple(statement.statements[:-1])),
+            statement.statements[-1],
+        )
+
+
+def construct_axiom(
+    antecedents: tuple[DisjointStatement | EssentialStatement, ...], consequent: ProvableStatement
+) -> AxiomaticStatement | Block:
+    if not antecedents:
+        return AxiomaticStatement(consequent.label, consequent.terms)
+    return Block((*antecedents, AxiomaticStatement(consequent.label, consequent.terms)))
 
-    for symbol in mm_symbols:
-        output_composer.load(ConstantStatement((symbol,)))
-    for variable in variables:
-        output_composer.load(VariableStatement((Metavariable(variable),)))
 
-    processed_floatings: list[FloatingStatement] = []
-    for floating in itertools.chain(*(lemma.context.floatings for lemma in (*lemmas, theorem))):
-        if floating in processed_floatings:
+def slice_database(input_database: Database) -> Iterator[tuple[str, Database]]:
+    """Of the top-level statements, only floating statements are mandatory hypothesis.
+    They are thus order sensitive.
+    """
+    cut_antecedents: dict[str, FloatingStatement | AxiomaticStatement | Block] = {}
+    global_disjoints: set[frozenset[str]] = set()
+
+    for statement in input_database.statements:
+        """Constants and variables can be ignored since we can deduce the set of constants from the parsed statement."""
+        if isinstance(statement, (ConstantStatement, VariableStatement)):
             continue
-        processed_floatings += [floating]
-        output_composer.load(floating)
-
-    for lemma in lemmas:
-        copy_with_context(
-            (*theorem_get_disjoints(lemma), *lemma.context.essentials),
-            AxiomaticStatement(lemma.statement.label, lemma.statement.terms),
-            output_composer,
-        )
-    copy_with_context(
-        (*theorem_get_disjoints(theorem), *theorem.context.essentials),
-        theorem.statement,
-        output_composer,
-    )
+        elif isinstance(statement, DisjointStatement):
+            for var1 in statement.metavariables:
+                for var2 in statement.metavariables:
+                    if var1 != var2:
+                        global_disjoints.add(frozenset({var1.name, var2.name}))
+        elif isinstance(statement, FloatingStatement):
+            cut_antecedents[statement.label] = statement
+        elif is_axiom(statement):
+            cut_antecedents[axiom_get_label(statement)] = cast('AxiomaticStatement | Block', statement)
+        elif isinstance(statement, (ProvableStatement, Block)):
+            antecedents, consequent = deconstruct_provable(statement)
+            yield (
+                consequent.label,
+                supporting_database_for_provable(cut_antecedents, global_disjoints, consequent, antecedents),
```

Comment:
Is the order guaranteed, i.e., the supporting database for the provable can only depend on `cut_antecedents`, what has been seen so far?